File: Exponent Solutions/Time Planner.py
from typing import List

def meeting_planner(slotsA: List[List[int]], slotsB: List[List[int]], dur: int) -> List[int]:
  # Two pointers
  a = b = 0
  A, B = len(slotsA), len(slotsB)
  while a < A and b < B:
    print(slotsA[a], slotsB[b])
    overlap, length = get_overlap(slotsA[a], slotsB[b])

    if overlap and length >= dur:
      return [overlap[0], overlap[0] + dur]

    else:
      # Check which one is behind
      if slotsA[a][1] <= slotsB[b][1]:
        # slotsA is behind
        a += 1
      elif slotsB[b][1] <= slotsA[a][1]:
        b += 1

  return []

def get_overlap(sA, sB):
  a_s, a_e = sA
  b_s, b_e = sB

  start = max(a_s, b_s)
  end = min(a_e, b_e)

  if start < end:
    return [start, end], end - start
  
  return None, None

File: Exponent Solutions/Time Planner.js
// Written on October 22, 2024 after seeing the solution
function meetingPlanner(slotsA, slotsB, dur) {
    // Two pointer approach
    let a = 0, b = 0;
    let A = slotsA.length, B = slotsB.length;
    let a_avail, b_avail;
    let start, end;

    while (a < A && b < B) {
        a_avail = slotsA[a];
        b_avail = slotsB[b];

        // get overlap
        start = Math.max(a_avail[0], b_avail[0]);
        end = Math.min(a_avail[1], b_avail[1]);

        if (start + dur <= end) 
            return [start, start + dur];  // found an answer

        // else, we need to move pointer to whichever one is behind forward 
        else if (a_avail[1] < b_avail[1])
            a++;
        else
            b++;
    }

    return [];
}
